File: R/weather_cmfd.R
# File: weather_cmfd.R   (R twin of python/dssatutils/weather_cmfd.py)
#
# Weather source: CMFD (China Meteorological Forcing Dataset) -> DSSAT .WTH.
#
# WHY: CMFD is the standard 0.1° near-surface forcing dataset for CHINA
# (1979-2018). It is 3-HOURLY, so this module aggregates each variable to the
# daily statistics DSSAT needs.
#
# CMFD variables (one NetCDF per variable per month; var name == token):
#   temp (K)  prec (mm/hr)  srad (W/m^2)  shum (kg/kg)  pres (Pa)  wind (m/s)
#
# Daily reduction:  TMAX/TMIN = max/min of temp-273.15;  RAIN = mean(prec)*24;
#   SRAD = mean(srad)*0.0864 MJ/m^2/day;  WIND = mean(wind);
#   RH/TDEW from daily-mean shum + temp + pres.
#
# ACCESS: free, but requires a (free) TPDC account (data.tpdc.ac.cn) to download
# the NetCDFs. Point cmfd_nc_dir at the folder. Coverage: China.

CMFD_TOKENS <- c("temp", "prec", "srad", "shum", "pres", "wind")

.cmfd_rh_tdew <- function(shum, temp_c, pres_pa) {
  q <- shum; p_hpa <- pres_pa / 100; tc <- temp_c
  e <- q * p_hpa / (0.622 + 0.378 * q)
  es <- 6.112 * exp(17.67 * tc / (tc + 243.5))
  rh <- pmin(pmax(100 * e / es, 0), 100)
  ln <- ifelse(e > 0, log(e / 6.112), NA_real_)
  tdew <- (243.5 * ln) / (17.67 - ln)
  list(rh = rh, tdew = tdew)
}

.cmfd_find_files <- function(nc_dir, token) {
  if (is.null(nc_dir) || !dir.exists(nc_dir)) return(character(0))
  files <- list.files(nc_dir, pattern = "\\.nc$", full.names = TRUE)
  base <- tolower(basename(files))
  files[startsWith(base, paste0(token, "_")) | grepl(paste0("_", token, "_"), base) |
        grepl(paste0(token, "-"), base)]
}

# Per-point daily reductions for one variable: returns {pid: data.frame(date,...)}.
.cmfd_extract_daily <- function(paths, token, ids, pts_vect) {
  acc <- setNames(lapply(ids, function(.) NULL), ids)   # per-pid list of (time,val)
  for (path in paths) {
    r <- terra::rast(path)
    tt <- terra::time(r)
    if (all(is.na(tt))) next
    ex <- terra::extract(r, pts_vect, ID = FALSE)        # points x timesteps
    for (j in seq_along(ids)) {
      df <- data.frame(t = as.POSIXct(tt), v = as.numeric(ex[j, ]))
      acc[[ids[j]]] <- rbind(acc[[ids[j]]], df)
    }
  }
  out <- setNames(vector("list", length(ids)), ids)
  for (pid in ids) {
    d <- acc[[pid]]
    if (is.null(d) || nrow(d) == 0) { out[[pid]] <- data.frame(); next }
    d <- d[!duplicated(d$t), ]
    day <- as.Date(d$t)
    agg <- switch(token,
      temp = data.frame(date = sort(unique(day)),
                        TMAX = tapply(d$v, day, max, na.rm = TRUE) - 273.15,
                        TMIN = tapply(d$v, day, min, na.rm = TRUE) - 273.15,
                        TMEAN = tapply(d$v, day, mean, na.rm = TRUE) - 273.15),
      prec = data.frame(date = sort(unique(day)), RAIN = tapply(d$v, day, mean, na.rm = TRUE) * 24),
      srad = data.frame(date = sort(unique(day)), SRAD = tapply(d$v, day, mean, na.rm = TRUE) * 0.0864),
      wind = data.frame(date = sort(unique(day)), WIND = tapply(d$v, day, mean, na.rm = TRUE)),
      shum = data.frame(date = sort(unique(day)), SHUM = tapply(d$v, day, mean, na.rm = TRUE)),
      pres = data.frame(date = sort(unique(day)), PRES = tapply(d$v, day, mean, na.rm = TRUE)),
      data.frame(date = sort(unique(day)), V = tapply(d$v, day, mean, na.rm = TRUE)))
    rownames(agg) <- NULL
    out[[pid]] <- agg
  }
  out
}

.cmfd_write_wth <- function(df, pid, lat, lon, output_dir) {
  tavg <- (df$TMAX + df$TMIN) / 2
  tav <- mean(tavg, na.rm = TRUE)
  mon <- tapply(tavg, list(df$YEAR, df$MM), mean, na.rm = TRUE)
  amps <- apply(mon, 1, function(r) { r <- r[is.finite(r)]; if (length(r)) max(r) - min(r) else NA })
  amp <- mean(amps, na.rm = TRUE)
  header <- sprintf(
    "$WEATHER DATA: CMFD (Point ID: %s)\n@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT\n  CMFD %8.4f %8.4f   -99 %5.1f %5.1f   2.0  10.0\n@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND",
    pid, lat, lon, tav, amp)
  clamp <- function(x) ifelse(!is.na(x) & (x >= 9999.95 | x <= -999.95), -99, x)
  d <- df; for (c in c("SRAD","TMAX","TMIN","RAIN","TDEW","RH2M","WIND")) d[[c]][is.na(d[[c]])] <- -99
  lines <- sprintf("%7s%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f",
                   d$DATE, clamp(d$SRAD), clamp(d$TMAX), clamp(d$TMIN),
                   clamp(d$RAIN), clamp(d$TDEW), clamp(d$RH2M), clamp(d$WIND))
  lines <- gsub("-99.0", "  -99", lines, fixed = TRUE)
  writeLines(c(header, lines), file.path(output_dir, sprintf("%s.WTH", pid)))
}

process_weather_cmfd <- function(shapefile, start_year, end_year, output_dir,
                                 id_col, lat_col, lon_col, n_cores, log_file,
                                 cmfd_nc_dir = "") {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  end_year <- min(end_year, as.integer(format(Sys.Date(), "%Y")))
  if (!nzchar(cmfd_nc_dir) || !dir.exists(cmfd_nc_dir))
    stop("CMFD needs cmfd_nc_dir pointing at a folder of CMFD NetCDF files ",
         "(temp/prec/srad/shum/pres/wind). Download from the TPDC (free account).")

  pts <- sf::st_transform(shapefile, 4326)
  ids <- as.character(sf::st_drop_geometry(pts)[[id_col]])
  xy <- sf::st_coordinates(pts); lats <- xy[, 2]; lons <- xy[, 1]
  pts_vect <- terra::vect(pts)

  message(sprintf("--- Starting CMFD Processing (Years: %d-%d) ---", start_year, end_year))
  var_files <- lapply(CMFD_TOKENS, function(tok) .cmfd_find_files(cmfd_nc_dir, tok))
  names(var_files) <- CMFD_TOKENS
  if (length(var_files$temp) == 0) stop("CMFD requires at least the temp NetCDF files.")

  daily <- list()
  for (tok in CMFD_TOKENS) if (length(var_files[[tok]]) > 0)
    daily[[tok]] <- .cmfd_extract_daily(var_files[[tok]], tok, ids, pts_vect)

  written <- 0
  for (k in seq_along(ids)) {
    pid <- ids[k]; lat <- lats[k]; lon <- lons[k]
    tryCatch({
      tdf <- daily[["temp"]][[pid]]
      if (is.null(tdf) || nrow(tdf) == 0) stop("No CMFD temperature data extracted (point outside China grid).")
      frame <- tdf
      jn <- function(tok) { d <- daily[[tok]][[pid]]; if (!is.null(d) && nrow(d)) merge(frame, d, by = "date", all.x = TRUE) else frame }
      for (tok in c("prec", "srad", "wind")) frame <- jn(tok)
      sh <- daily[["shum"]][[pid]]; pr <- daily[["pres"]][[pid]]
      if (!is.null(sh) && nrow(sh) && !is.null(pr) && nrow(pr)) {
        m <- merge(merge(frame[, c("date", "TMEAN")], sh, by = "date", all.x = TRUE), pr, by = "date", all.x = TRUE)
        rt <- .cmfd_rh_tdew(m$SHUM, m$TMEAN, m$PRES)
        frame <- merge(frame, data.frame(date = m$date, RH2M = rt$rh, TDEW = rt$tdew), by = "date", all.x = TRUE)
      }
      yr <- as.integer(format(frame$date, "%Y"))
      frame <- frame[yr >= start_year & yr <= end_year, ]
      if (nrow(frame) == 0) stop("No CMFD data in the requested year range.")
      frame$DATE <- sprintf("%d%03d", as.integer(format(frame$date, "%Y")), as.integer(format(frame$date, "%j")))
      frame$YEAR <- as.integer(format(frame$date, "%Y")); frame$MM <- as.integer(format(frame$date, "%m"))
      for (c in c("SRAD","RAIN","RH2M","WIND","TDEW")) if (is.null(frame[[c]])) frame[[c]] <- -99
      frame <- frame[!is.na(frame$TMAX) & !is.na(frame$TMIN), ]
      .cmfd_write_wth(frame, pid, lat, lon, output_dir)
      written <- written + 1
    }, error = function(e) {
      msg <- sprintf("\n--- ERROR ---\nCMFD point %s (%.3f,%.3f): %s\n", pid, lat, lon, conditionMessage(e))
      cat(msg); write(msg, file = log_file, append = TRUE)
    })
  }
  message(sprintf("\nCMFD processing complete: %d/%d points written to '%s'.\n", written, length(ids), output_dir))
}

File: python/dssatutils/weather_cmfd.py
# File: weather_cmfd.py
# Python port of weather_cmfd.R
#
# Weather source: CMFD (China Meteorological Forcing Dataset) -> DSSAT .WTH.
#
# WHY: CMFD is the standard high-resolution (0.1°) near-surface forcing dataset
# for CHINA, 1979-2018, fusing station, reanalysis and satellite data. It is
# 3-HOURLY, so this module aggregates each variable to the daily statistics DSSAT
# needs (Tmax/Tmin from sub-daily temperature, daily precip total, mean solar
# radiation, etc.).
#
# CMFD variables (one NetCDF per variable per month; var name == token):
#   temp (K, 2 m air T)   prec (mm/hr, precip rate)   srad (W/m^2, dwn shortwave)
#   shum (kg/kg, specific humidity)   pres (Pa)   wind (m/s, 10 m)
#
# Daily reduction:  TMAX/TMIN = max/min of temp-273.15;  RAIN = mean(prec)*24;
#   SRAD = mean(srad)*0.0864 MJ/m^2/day;  WIND = mean(wind);
#   RH/TDEW from daily-mean shum + temp + pres.
#
# ACCESS: free, but requires a (free) account at the National Tibetan Plateau
# Data Center (data.tpdc.ac.cn) to DOWNLOAD the NetCDF files. Point *cmfd_nc_dir*
# at the downloaded folder. Coverage: China.

import glob
import os
from datetime import date

import numpy as np
import pandas as pd

# DSSAT-relevant CMFD variables + filename/variable token.
_CMFD_TOKENS = {"temp": "temp", "prec": "prec", "srad": "srad",
                "shum": "shum", "pres": "pres", "wind": "wind"}


def _calc_tav(df):
    return float(((df["TMAX"] + df["TMIN"]) / 2.0).mean())


def _calc_amp(df):
    d = df.copy()
    d["TAVG"] = (d["TMAX"] + d["TMIN"]) / 2.0
    monthly = d.groupby(["YEAR", "MM"])["TAVG"].mean().reset_index()
    annual = monthly.groupby("YEAR")["TAVG"].agg(lambda x: x.max() - x.min())
    return float(annual.mean())


def _rh_tdew_from_shum(shum, temp_c, pres_pa):
    """Relative humidity (%) and dew point (°C) from specific humidity."""
    q = np.asarray(shum, dtype=float)
    p_hpa = np.asarray(pres_pa, dtype=float) / 100.0
    tc = np.asarray(temp_c, dtype=float)
    with np.errstate(invalid="ignore", divide="ignore"):
        e = q * p_hpa / (0.622 + 0.378 * q)            # actual vapour pressure (hPa)
        es = 6.112 * np.exp(17.67 * tc / (tc + 243.5))  # saturation VP (hPa)
        rh = np.clip(100.0 * e / es, 0.0, 100.0)
        ln = np.log(np.where(e > 0, e / 6.112, np.nan))
        tdew = (243.5 * ln) / (17.67 - ln)
    return rh, tdew


def _find_var_files(nc_dir, token):
    if not nc_dir or not os.path.isdir(nc_dir):
        return []
    hits = []
    for f in sorted(glob.glob(os.path.join(nc_dir, "*.nc"))):
        base = os.path.basename(f).lower()
        if base.startswith(f"{token}_") or f"_{token}_" in base or f"{token}-" in base:
            hits.append(f)
    return hits


def _extract_daily(paths, token, ids, lats, lons):
    """Return {pid: DataFrame(date-indexed daily reductions)} for one variable."""
    import xarray as xr
    parts = {pid: [] for pid in ids}
    pts_lat = xr.DataArray(np.asarray(lats), dims="points")
    pts_lon = xr.DataArray(np.asarray(lons), dims="points")
    for path in paths:
        with xr.open_dataset(path) as ds:
            ds = ds.load()
            dv = next((v for v in ds.data_vars), None)
            if dv is None or "time" not in ds.coords:
                continue
            latname = "latitude" if "latitude" in ds.coords else ("lat" if "lat" in ds.coords else None)
            lonname = "longitude" if "longitude" in ds.coords else ("lon" if "lon" in ds.coords else None)
            if latname is None or lonname is None:
                continue
            times = pd.to_datetime(ds["time"].values)
            sel = ds[dv].sel({latname: pts_lat, lonname: pts_lon}, method="nearest")
            vals = np.asarray(sel.values)
            if vals.ndim == 1:
                vals = vals.reshape(len(times), 1) if vals.shape[0] == len(times) else vals.reshape(1, -1).T
            elif vals.shape[0] != len(times):
                vals = vals.T
            for j, pid in enumerate(ids):
                s = pd.Series(vals[:, j], index=times)
                parts[pid].append(s)
    out = {}
    for pid in ids:
        if not parts[pid]:
            out[pid] = pd.DataFrame()
            continue
        s = pd.concat(parts[pid]).sort_index()
        s = s[~s.index.duplicated()]
        g = s.groupby(s.index.normalize())
        if token == "temp":                       # K -> daily max/min/mean (°C)
            daily = pd.DataFrame({"TMAX": g.max() - 273.15, "TMIN": g.min() - 273.15,
                                  "TMEAN": g.mean() - 273.15})
        elif token == "prec":                     # mm/hr -> daily total mm
            daily = pd.DataFrame({"RAIN": g.mean() * 24.0})
        elif token == "srad":                     # W/m^2 -> MJ/m^2/day
            daily = pd.DataFrame({"SRAD": g.mean() * 0.0864})
        elif token == "wind":
            daily = pd.DataFrame({"WIND": g.mean()})
        elif token == "shum":
            daily = pd.DataFrame({"SHUM": g.mean()})
        elif token == "pres":
            daily = pd.DataFrame({"PRES": g.mean()})
        else:
            daily = pd.DataFrame({token.upper(): g.mean()})
        out[pid] = daily
    return out


def _write_wth(df, pid, lat, lon, output_dir):
    tav = _calc_tav(df)
    amp = _calc_amp(df)
    header = (
        f"$WEATHER DATA: CMFD (Point ID: {pid})\n"
        f"@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT\n"
        f"  CMFD {lat:8.4f} {lon:8.4f}   -99 {tav:5.1f} {amp:5.1f}   2.0  10.0\n"
        f"@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND"
    )
    lines = []
    for _, row in df.iterrows():
        line = (
            f"{row['DATE']:>7s}"
            f"{row['SRAD']:6.1f}{row['TMAX']:6.1f}{row['TMIN']:6.1f}"
            f"{row['RAIN']:6.1f}{row['TDEW']:6.1f}{row['RH2M']:6.1f}{row['WIND']:6.1f}"
        )
        line = line.replace(" -99.0", "   -99")
        lines.append(line)
    out_path = os.path.join(output_dir, f"{pid}.WTH")
    with open(out_path, "w") as fh:
        fh.write(header + "\n")
        fh.write("\n".join(lines) + "\n")
    return out_path


def process_weather_cmfd(
    shapefile, start_year, end_year, output_dir,
    id_col, lat_col, lon_col, n_cores, log_file,
    cmfd_nc_dir="",
):
    """Build DSSAT .WTH from CMFD 3-hourly grids (aggregated to daily) per point.

    Point *cmfd_nc_dir* at the folder of pre-downloaded CMFD NetCDFs
    (temp/prec/srad/shum/pres/wind). temp is required; RH/TDEW are derived from
    specific humidity + pressure + temperature when those files are present.
    """
    os.makedirs(output_dir, exist_ok=True)
    end_year = min(end_year, date.today().year)
    if not cmfd_nc_dir or not os.path.isdir(cmfd_nc_dir):
        raise FileNotFoundError(
            "CMFD needs cmfd_nc_dir pointing at a folder of CMFD NetCDF files "
            "(temp/prec/srad/shum/pres/wind). Download from the TPDC (free account).")

    pts = shapefile.copy()
    if hasattr(pts, "geometry"):
        pts = pts.to_crs("EPSG:4326")
        pts[lat_col] = pts.geometry.y
        pts[lon_col] = pts.geometry.x
    ids = [str(r[id_col]) for _, r in pts.iterrows()]
    lats = np.array([float(r[lat_col]) for _, r in pts.iterrows()])
    lons = np.array([float(r[lon_col]) for _, r in pts.iterrows()])

    print(f"--- Starting CMFD Processing (Years: {start_year}-{end_year}) ---")
    var_files = {tok: _find_var_files(cmfd_nc_dir, tok) for tok in _CMFD_TOKENS}
    if not var_files["temp"]:
        raise FileNotFoundError("CMFD requires at least the temp NetCDF files.")

    daily = {tok: _extract_daily(paths, tok, ids, lats, lons)
             for tok, paths in var_files.items() if paths}

    written = 0
    for pid, lat, lon in zip(ids, lats, lons):
        try:
            tdf = daily.get("temp", {}).get(pid, pd.DataFrame())
            if tdf.empty:
                raise ValueError("No CMFD temperature data extracted (point outside the China grid).")
            frame = tdf.copy()
            for tok, col in (("prec", "RAIN"), ("srad", "SRAD"), ("wind", "WIND")):
                d = daily.get(tok, {}).get(pid)
                frame = frame.join(d[[col]], how="left") if (d is not None and not d.empty) else frame
            # RH / TDEW from specific humidity + pressure + daily-mean temp.
            shum = daily.get("shum", {}).get(pid)
            pres = daily.get("pres", {}).get(pid)
            if shum is not None and not shum.empty and pres is not None and not pres.empty:
                j = frame.join(shum, how="left").join(pres, how="left")
                rh, tdew = _rh_tdew_from_shum(j["SHUM"].values, j["TMEAN"].values, j["PRES"].values)
                frame["RH2M"] = rh
                frame["TDEW"] = tdew
            frame = frame.reset_index().rename(columns={"index": "DT"})
            frame["DT"] = pd.to_datetime(frame["DT"])
            frame = frame[(frame["DT"].dt.year >= start_year) & (frame["DT"].dt.year <= end_year)]
            if frame.empty:
                raise ValueError("No CMFD data in the requested year range.")
            frame["DATE"] = frame["DT"].apply(lambda t: f"{t.year}{t.dayofyear:03d}")
            frame["YEAR"] = frame["DT"].dt.year
            frame["MM"] = frame["DT"].dt.month
            for need in ("SRAD", "RAIN", "RH2M", "WIND", "TDEW"):
                if need not in frame:
                    frame[need] = -99.0
            frame = frame[frame["TMAX"].notna() & frame["TMIN"].notna()].fillna(-99)
            _write_wth(frame, pid, lat, lon, output_dir)
            written += 1
        except Exception as exc:  # noqa: BLE001
            msg = f"\n--- ERROR ---\nCMFD point {pid} ({lat:.3f},{lon:.3f}): {exc}\n"
            print(msg)
            with open(log_file, "a") as lf:
                lf.write(msg)
    print(f"\nCMFD processing complete: {written}/{len(ids)} points written to '{output_dir}'.\n")
